File: Top 100 Codes/01-Getting Started/15-Armstrong number in a given range/Armstrongnumberinagivenrange.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Function to check if a number is an Armstrong number
bool isArmstrong(int num) {
    int originalNum = num; // Store the original number
    int sum = 0;
    int digits = 0;

    // Calculate the number of digits
    int temp = num;
    while (temp != 0) {
        temp /= 10;
        digits++;
    }

    // Calculate the sum of each digit raised to the power of the number of digits
    temp = num;
    while (temp != 0) {
        int rem = temp % 10;        // Extract the last digit
        sum += pow(rem, digits);    // Add the power of the digit to the sum
        temp /= 10;                 // Remove the last digit
    }

    // Check if the sum is equal to the original number
    return sum == originalNum;
}

int main() {
    int start, end;

    // Prompt the user for the range
    cout << "Enter the start of the range: ";
    cin >> start;
    cout << "Enter the end of the range: ";
    cin >> end;

    // Validate the range
    if (start > end) {
        cout << "Invalid range! The start must be less than or equal to the end." << endl;
        return 1;
    }

    cout << "Armstrong numbers between " << start << " and " << end << " are:" << endl;

    // Loop through the range and check for Armstrong numbers
    bool found = false;
    for (int i = start; i <= end; i++) {
        if (isArmstrong(i)) {
            cout << i << " ";
            found = true;
        }
    }

    // If no Armstrong numbers are found
    if (!found) {
        cout << "None" << endl;
    } else {
        cout << endl;
    }

    return 0;
}

File: Top 100 Codes/01-Getting Started/15-Armstrong number in a given range/Armstrongnumberinagivenrange.py
# Function to check if a number is an Armstrong number
def is_armstrong(num):
    """
    Checks whether the given number is an Armstrong number.
    An Armstrong number is a number that is equal to the sum of its digits
    raised to the power of the number of digits.
    """
    num_str = str(num)  # Convert the number to a string
    num_digits = len(num_str)  # Find the number of digits
    armstrong_sum = sum(int(digit) ** num_digits for digit in num_str)  # Sum of powered digits
    return armstrong_sum == num  # Check if sum equals the original number

# Prompt the user for the range
start = int(input("Enter the start of the range: "))
end = int(input("Enter the end of the range: "))

# Validate the range
if start > end:
    print("Invalid range! The start must be less than or equal to the end.")
else:
    print(f"Armstrong numbers between {start} and {end} are:")

    # Find and print all Armstrong numbers in the range
    found = False
    for num in range(start, end + 1):
        if is_armstrong(num):
            print(num, end=" ")
            found = True

    # If no Armstrong numbers are found
    if not found:
        print("None")
    else:
        print()  # Print a newline for better formatting
